core/sr-api/src/lib.rs

```diff
@@ -0,0 +1,589 @@
+			$( #[$attr] )*
+			pub trait $name $(< $( $generic_param_orig $( : $generic_bound_orig )* ),* >)* {
+				$(
+					fn $fn_name $( < $( $fn_generic ),* > )* (
+						$( $param_name : $param_type )*
+					) $( -> $return_ty )*;
+				)*
+			};
+			$( $generic_param_parsed $( : $generic_bound_parsed )* ),*;
+			{ $( $result_return_ty; )* };
+			$( $( $generic_param_orig $( : $generic_bound_orig )*, )* )* $( $( $( $fn_generic, )* )* )*;
+		);
+	};
+	(@GENERATE_CLIENT_TRAITS
+		$( #[$attr:meta] )*
+		pub trait $name:ident $(< $( $generic_param_orig:ident $( : $generic_bound_orig:ident )* ),* >)* {
+			$(
+				fn $fn_name:ident $( < $( $fn_generic:ident ),* > )* (
+					$( $param_name:ident : $param_type:ty )*
+				) $( -> $return_ty:ty)*;
+			)*
+		};
+		$( $generic_param_parsed:ident $( : $generic_bound_parsed:path )* ),*;
+		{ $( $result_return_ty:ty; )* };
+		$( $generic_param_joined:ident $( : $generic_bound_joined:ident )*, )*;
+	) => {
+		$( #[$attr] )*
+		pub trait $name < $( $generic_param_parsed $( : $generic_bound_parsed )* ),* > {
+			type Error;
+
+			$(
+				fn $fn_name $( < $( $fn_generic: $crate::Encode + $crate::Decode ),* > )* (
+					&self, at: &$crate::BlockId<Block> $(, $param_name: $param_type )*
+				) -> $result_return_ty;
+			)*
+		}
+	};
+	(@GENERATE_RUNTIME_TRAITS
+		$(
+			$( #[$attr:meta] )*
+			pub trait $name:ident $(< $( $generic_param:ident $( : $generic_bound:ident )* ),* >)* {
+				$(
+					fn $fn_name:ident $( < $( $fn_generic:ident ),* > )* (
+						$( $param_name:ident : $param_type:ty )*
+					) $( -> $return_ty:ty)*;
+				)*
+			};
+		)*
+	) => {
+		decl_apis! {
+			@GENERATE_RUNTIME_TRAITS_WITH_JOINED_GENERICS
+			$(
+				$( #[$attr] )*
+				pub trait $name < $( $( $generic_param $( : $generic_bound )*, )* )* $( $( $( $fn_generic, )* )* )* > {
+					$(
+						fn $fn_name ($( $param_name: $param_type ),*) $( -> $return_ty )*;
+					)*
+				}
+			)*
+		}
+	};
+	(@GENERATE_RUNTIME_TRAITS_WITH_JOINED_GENERICS
+		$(
+			$( #[$attr:meta] )*
+			pub trait $name:ident < $( $generic_param:ident $( : $generic_bound:ident )*, )* > {
+				$(
+					fn $fn_name:ident($( $param_name:ident : $param_type:ty )*) $( -> $return_ty:ty)*;
+				)*
+			}
+		)*
+	) => {
+		pub mod runtime {
+			use super::*;
+
+			$(
+				$( #[$attr] )*
+				pub trait $name < $( $generic_param $( : $generic_bound )* ),* > {
+					$(
+						fn $fn_name ($( $param_name: $param_type ),*) $( -> $return_ty )*;
+					)*
+				}
+			)*
+		}
+	};
+}
+
+decl_apis! {
+	/// The `Core` api trait that is mandantory for each runtime.
+	pub trait Core<Block: BlockT, AuthorityId> {
+		fn version() -> RuntimeVersion;
+		fn authorities() -> Vec<AuthorityId>;
+		fn execute_block(block: Block);
+	}
+
+	/// The `Metadata` api trait that returns metadata for the runtime.
+	pub trait Metadata {
+		fn metadata() -> Vec<u8>;
+	}
+
+	/// The `BlockBuilder` api trait that provides required functions for building a block for a runtime.
+	pub trait BlockBuilder<Block: BlockT> {
```

Comment:
maybe move this closer to the actual block builder logic.

Reply:
I moved this into substrate client to the block builder, but now after being almost finished. I realized that the client is not build for wasm yet. Should I make the client compile on wasm or should I put the trait somewhere else?